File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        d = {")":"(", "}":"{", "]":"["}

        for val in s:
            if val in d:  # if it is a closing brcaket
                if stack and stack[-1] == d[val]:
                    stack.pop()
                else:
                    return False


            else:
                stack.append(val)

        return True if not stack else False
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> pairs = new HashMap<>();
        pairs.put(')', '(');
        pairs.put('}', '{');
        pairs.put(']', '[');

        Deque<Character> stack = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            if (pairs.containsKey(c)) {                 // closing bracket
                if (!stack.isEmpty() && stack.peek() == pairs.get(c)) {
                    stack.pop();
                } else {
                    return false;
                }
            } else {                                     // opening bracket
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}
